### PR TITLE
move apps storage to the apps group

### PR DESCRIPTION
**Pull request by deads2k**
This moves the build REST storage into the build package for clean separation, maintenance, and composition.

### CODE REVIEW

File: pkg/cmd/server/origin/openshift_apiserver.go [hunk 1]
```diff
@@ -189,6 +191,12 @@ type legacyStorageMutator interface {
 	mutate(map[schema.GroupVersion]map[string]rest.Storage)
 }
 
+type legacyStorageMutatorFunc func(map[schema.GroupVersion]map[string]rest.Storage)
```
Comment (mfojtik) [inline pkg/cmd/server/origin/openshift_apiserver.go]: maybe worth a little godoc explaining why is this needed and if/when this will go away?

Comment (mfojtik) [reply pkg/cmd/server/origin/openshift_apiserver.go]: also is this apps only or do we need this in other api servers?

File: pkg/deploy/apiserver/apiserver.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	osclient "github.com/openshift/origin/pkg/client"
+	appsapiv1 "github.com/openshift/origin/pkg/deploy/apis/apps/v1"
+	oappsclient "github.com/openshift/origin/pkg/deploy/generated/internalclientset"
+	deployconfigetcd "github.com/openshift/origin/pkg/deploy/registry/deployconfig/etcd"
+	deploylogregistry "github.com/openshift/origin/pkg/deploy/registry/deploylog"
+	deployconfiginstantiate "github.com/openshift/origin/pkg/deploy/registry/instantiate"
+	deployrollback "github.com/openshift/origin/pkg/deploy/registry/rollback"
+)
+
+type AppsConfig struct {
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: godoc?

File: pkg/deploy/apiserver/apiserver.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	*AppsConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *AppsConfig) Complete() completedConfig {
+	c.GenericConfig.Complete()
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: do we need to check for any errors here (in case we are unable to complete the config (if such case exists)) ?

File: pkg/deploy/apiserver/apiserver.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// SkipComplete provides a way to construct a server instance without config completion.
+func (c *AppsConfig) SkipComplete() completedConfig {
+	return completedConfig{c}
+}
+
+// New returns a new instance of AppsServer from the given config.
+func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*AppsServer, error) {
+	genericServer, err := c.AppsConfig.GenericConfig.SkipComplete().New("apps.openshift.io-apiserver", delegationTarget) // completion is done in Complete, no need for a second time
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: move the comment on top

Comment (mfojtik) [reply pkg/deploy/apiserver/apiserver.go]: also `apps.openshift.io-apiserver` looks like const?

Comment (mfojtik) [reply pkg/deploy/apiserver/apiserver.go]: also is `io-apiserver` some new pattern or why we can't use `apps.openshift.io/apiserver` ?

File: pkg/deploy/apiserver/apiserver.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	kubeInternalClient, err := kclientsetinternal.NewForConfig(c.CoreAPIServerClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeExternalClient, err := kclientsetexternal.NewForConfig(c.CoreAPIServerClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	//TODO/REBASE use something other than c.KubeClientsetInternal
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: doesn't seems to be valid anymore?

File: pkg/deploy/apiserver/apiserver.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	v1Storage["deploymentConfigs"] = deployConfigStorage
+	v1Storage["deploymentConfigs/scale"] = deployConfigScaleStorage
+	v1Storage["deploymentConfigs/status"] = deployConfigStatusStorage
+	v1Storage["deploymentConfigs/rollback"] = deployConfigRollbackStorage
+	v1Storage["deploymentConfigs/log"] = deploylogregistry.NewREST(deprecatedOpenshiftClient, kubeInternalClient.Core(), kubeInternalClient.Core(), nodeConnectionInfoGetter)
+	v1Storage["deploymentConfigs/instantiate"] = dcInstantiateStorage
+	return v1Storage, nil
+}
+
+// this allows us to inject a one-off endpoint into oapi
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: LegacyLegacyDCRollbackMutator allows us...

File: pkg/deploy/apiserver/apiserver.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+type LegacyLegacyDCRollbackMutator struct {
+	CoreAPIServerClientConfig *restclient.Config
+	Version                   schema.GroupVersion
+}
+
+func (l LegacyLegacyDCRollbackMutator) Mutate(legacyStorage map[schema.GroupVersion]map[string]rest.Storage) {
+	externalVersionCodec := kapi.Codecs.LegacyCodec(schema.GroupVersion{Group: "", Version: "v1"})
+
+	originAppsClient, err := oappsclient.NewForConfig(l.CoreAPIServerClientConfig)
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: NewForConfigOrDie ?

File: pkg/deploy/apiserver/apiserver.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		panic(err)
+	}
+	kubeInternalClient, err := kclientsetinternal.NewForConfig(l.CoreAPIServerClientConfig)
```
Comment (mfojtik) [inline pkg/deploy/apiserver/apiserver.go]: NewForConfigOrDie ?

File: pkg/cmd/server/origin/openshift_apiserver.go (hunk 1 above)
Comment (deads2k) [reply pkg/cmd/server/origin/openshift_apiserver.go]: > maybe worth a little godoc explaining why is this needed and if/when this will go away?

I'm not sure of the final factorization. This pattern began a couple apiservers back.  `users` I think.

Comment (deads2k) [reply pkg/cmd/server/origin/openshift_apiserver.go]: > also is this apps only or do we need this in other api servers?

The interface is doced

File: pkg/deploy/apiserver/apiserver.go (hunk 3 above)
Comment (deads2k) [reply pkg/deploy/apiserver/apiserver.go]: > do we need to check for any errors here (in case we are unable to complete the config (if such case exists)) ?

`Complete` doesn't return an error.

**Comment by deads2k** [comment]
comments addressed
